Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1486,101 +1487,196 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
   return result;
 }
 
-// TODO(varconst): everything below
-
 MutationResult Serializer::DecodeMutationResult(
+    nanopb::Reader* reader,
     const google_firestore_v1_WriteResult& write_result,
     const SnapshotVersion& commit_version) const {
-  // TODO(varconst): implement.
-  // - (MutationResult)decodedMutationResult:(GCFSWriteResult *)mutation
-  //     commitVersion:(const SnapshotVersion &)commitVersion;
-  return MutationResult{{}, {}};
+  // NOTE: Deletes don't have an update_time, use commit_version instead.
+  SnapshotVersion version =
+      write_result.has_update_time
+          ? DecodeSnapshotVersion(reader, write_result.update_time)
+          : commit_version;
+
+  absl::optional<std::vector<FieldValue>> transform_results;
+  if (write_result.transform_results_count > 0) {
+    transform_results = std::vector<FieldValue>{};
+    for (pb_size_t i = 0; i < write_result.transform_results_count; i++) {
+      transform_results->push_back(
+          DecodeFieldValue(reader, write_result.transform_results[i]));
+    }
+  }
+
+  return MutationResult(version, std::move(transform_results));
 }
 
 std::unordered_map<std::string, std::string>
 Serializer::EncodeListenRequestLabels(const QueryData& query_data) const {
-  // TODO(varconst): implement.
-  // - (nullable NSMutableDictionary<NSString *, NSString *> *)
-  //     encodedListenRequestLabelsForQueryData: (const QueryData &)queryData;
-  return {};
+  auto value = EncodeLabel(query_data.purpose());
```

Comment:
This return type is wrong: the serializer needs to produce an array of `google_firestore_v1_ListenRequest_LabelsEntry`, not a `std::unordered_map<std::string, std::string>`.

Reply:
Done.